lang/python/CMakeLists.txt

```diff
@@ -23,17 +23,30 @@ include(${SWIG_USE_FILE})
 
 set(python_libs)
 set(python_version)
+
+set(required_version)
+if(PYTHON3_REQUIRED_VERSION)
+    if(NOT PYTHON3_REQUIRED_VERSION MATCHES "^([0-9]+)(\\.[0-9]+(\\.[0-9]+)?)?$")
+        message(FATAL_ERROR "Invalid value for PYTHON3_REQUIRED_VERSION: Requires a valid version string \
+            Provide a version number following the format: major[.minor[.patch]]")
+    endif()
+    if ("${PYTHON3_REQUIRED_VERSION}" VERSION_LESS 3)
+        message(FATAL_ERROR "Invalid value for PYTHON3_REQUIRED_VERSION: Requires a Python version >= 3")
+    endif()
+    set(required_version ${PYTHON3_REQUIRED_VERSION} EXACT)
+endif()
+
 if("${CMAKE_VERSION}" VERSION_LESS "3.12.0")
     # This method of finding python libs has been deprecated since version 3.12.
     # If we are running with a greater CMake version, opt to use the Python3 package.
     set(Python_ADDITIONAL_VERSIONS 3.9 3.8 3.7 3.6 3.5)
-    find_package(PythonInterp REQUIRED)
-    find_package(PythonLibs REQUIRED)
+    find_package(PythonInterp ${required_version} REQUIRED)
```

Comment:
Just for my understanding: if a user does not specify the new `PYTHON3_REQUIRED_VERSION` config option, I assume `${required_version}` would be left as unset/undefined. Will `find_package` here fail since `REQUIRED` is specified?

Reply:
Good question. `find_package` won't fail. 
Given that `${required_version}` is unset,  `${required_version}` gets substituted as an empty string in the expression when evaluating `find_package(PythonInterp ${required_version} REQUIRED)`.  This then just becomes `find_package(PythonInterp REQUIRED)`, as how it was prior to this change. 
This being the the default behaviour, taking that highest python 3 version it can find.

The `REQUIRED` keyword at the end of the expression is telling `find_package` to fail if it can't find any python package. If `PYTHON3_REQUIRED_VERSION` is set, that would mean it can't find that specific version. If `PYTHON3_REQUIRED_VERSION` unset, that would mean it can't find any python3 version.

Reply:
Thanks for the answers, which are very helpful 👍